### PR TITLE
Support for ios17

### PR DESCRIPTION
**Pull request by kali**

### CODE REVIEW

**Comment by kali** [comment]
Ping @simlay here... I'd like to get your opinion on this. It can be improved a bit (not starting the tunnel and devserver at each run maybe), but it feels super fragile to me at this stage.

Also, from the look of it, you're in a position to test on some variety of devices (I just could just try my iphone12/ios17). And I don't know if the new tunnel workflow will work with older devices or if it is a brand new thing. If it only works with ios17, we need to preserve the old workflow and use the detected ios version to enable the right one.

 What do you think ?

**Comment by simlay** [comment]
Nicely done. After updating my version of `pymobiledevice3`, this worked well for iOS 17 on my iPhone 15. `xcrun devicectl` (such as `list devices`) doesn't work with anything less than iOS 17 which makes this more annoying.  That's why I added the `version` to the `IosDevice` struct and kept using `ios-deploy -c` to detect devices. The `UUID` from `ios-deploy` works with `xcrun devicectl` (unclear about `pymobiledevice3`). 

Couple notes:
* Before the sudo password is prompted, maybe say why and what commands will be ran.
* When the app is running and I `ctrl-c`, it looks like the `lldb` command remains running even if the app on the device is closed. I just see it in `ps`: `96252 ttys003    0:05.09 /Applications/Xcode.app/Contents/Developer/usr/bin/lldb -s /var/folders/dq/kjbf14qn4zx3qw_3gtpvmjr40000gn/T/dinghy-lldb.Mtjyy0elxkDg/run.lldb` Not sure if there's a great way to deal with that. Just something to keep in mind.

**Comment by kali** [comment]
Thanks for checking.

* Ok, I will re-instantiate the ios-deploy stuff beside the new workflow.
* Agreed on the sudo prompt message.
* I'm not happy with all the processes I have to spawn and do not really join correctly in case of errors. And I suspect leaving them in the background participates to at least one of the failure mode after a dozen of runs or so. I'll try to make it better.

File: dinghy-lib/src/apple/device.rs [hunk 1]
```diff
@@ -37,31 +42,44 @@ pub struct AppleSimDevice {
...
+    fn is_locked(&self) -> Result<bool> {
+        let result = process::Command::new("xcrun")
+            .args(
+                "devicectl device info lockState --quiet --json-output /dev/stdout --device"
+                    .split_whitespace(),
+            )
+            .arg(&self.id)
+            .log_invocation(1)
+            .output()
+            .context("Failed to run ios-deploy")?;
```
Comment (simlay) [inline dinghy-lib/src/apple/device.rs:72]:
```suggestion
            .context("Failed to run devicectl device info lockState")?;
```

File: dinghy-lib/src/apple/device.rs [hunk 2]
```diff
@@ -91,24 +109,155 @@ impl IosDevice {
...
+        }
+
+        // xcrun devicectl device install app --device 00008110-001XXXXXXXXXX ./xgen/Build/Products/Release-iphoneos/nilo.app
+        let result = process::Command::new("xcrun")
+            .args("devicectl device install app --device".split_whitespace())
+            .arg(&self.id)
+            .arg(&*bundle)
+            .log_invocation(1)
+            .status()
+            .context("Failed to run ios-deploy")?;
```
Comment (simlay) [inline dinghy-lib/src/apple/device.rs:129]:
```suggestion
            .context("Failed to run devicectl device install app")?;
```

File: dinghy-lib/src/apple/mod.rs [hunk 3]
```diff
@@ -120,40 +122,35 @@ impl PlatformManager for WatchosManager {
...
+                } else if *arch == "x86_64-sim" {
+                    format!("x86_64-apple-watchos-sim")
+                } else {
+                    format!("{}-apple-watchos", arch)
+                };
+                let simulator = if *arch == "x86_64-sim" || *arch == "aarch64-sim" {
+                    Some(AppleSimulatorType::Watchos)
+                } else {
+                    None
+                };
```
Comment (simlay) [inline dinghy-lib/src/apple/mod.rs:143]: This from `cargo fmt`?

File: docs/ios.md [hunk 4]
```diff
@@ -1,21 +1,20 @@
...
 
-### Additional Requirements
+### Additional iOS Requirements
+
+You will need:
 
-In order to deploy to iOS devices one needs to install [ios-deploy](https://github.com/ios-control/ios-deploy). For example with:
-`brew install ios-deploy`
+ * XCode (the full thing, not restricted to the command line)
+ * `pymobiledevice3` for deploying and running code on the devices. `pip install pymobiledevice3` should do the trick.
```
Comment (simlay) [inline docs/ios.md:16]:
```suggestion
 * [`pymobiledevice3`](https://github.com/doronz88/pymobiledevice3) for deploying and running code on the devices. `pip install pymobiledevice3` should do the trick.
```

**Review by simlay — APPROVED** [review]
Looks pretty good! I saw some minor typos.

I tested this on:
* iPhone 7 - iOS 15.8.1
* iPhone 8 - iOS 16.7.5
* iPhone 15 - iOS 17.3.1

iOS 15 and 16 worked when running over wifi (via `ios-deploy`) and with lightning cable tether.

~Not sure how much you care but iOS 17 (iPhone 15) only worked via cabled attached. When trying via wifi, `devicectl device install app --device ...` succeeded.~

~The tunneling part failed with:~

```
Please enter simlay's password on m1-mbp-si to start a tunnel to 'iphone-15-si' (sudo):
2024-02-28 12:32:48 m1-mbp-si pymobiledevice3.__main__[38342] ERROR Device is not connected
[2024-02-28T17:32:48Z DEBUG dinghy_lib::apple::device] iOS RSD tunnel started:
Error: Option '--rsd' requires 2 arguments.
thread 'main' panicked at dinghy-lib/src/apple/device.rs:184:14:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
[2024-02-28T17:32:49Z DEBUG cargo_dinghy] done
```

pymobiledevice3 doesn't currently support tunneling via wifi. https://github.com/doronz88/pymobiledevice3/issues/634

File: dinghy-lib/src/apple/mod.rs (hunk 3 above)
Comment (kali) [reply dinghy-lib/src/apple/mod.rs:143]: I... guess so :)

**Comment by kali** [comment]
Thanks for reviewing despite the annoying formatting fixes.

How much do we want untethered ? We could maybe have it working partially with devicectl, but without capture of course. It may come "free" with a bump of pymobiledevice3 later. We — sonos — don't care that much about untethered. Do you ?

**Comment by simlay** [comment]
> How much do we want untethered ? We could maybe have it working partially with devicectl, but without capture of course. It may come "free" with a bump of pymobiledevice3 later. We — sonos — don't care that much about untethered. Do you ?

Yeah... I notice that untethered stuff is less reliable using just the `xcrun devicectl` tooling. Thanks for the work!

**Comment by kali** [comment]
ok. I'm gonna merge this as is then. I think the best way to get wifi would actually be to contribute to pymobiledevice, assuming it can be made to work. I am very reluctant to introduce workflows where we can not get stdout and the return value as running test and bench is historically the core feature of dinghy...

**Pull request merged**